### PR TITLE
Dynamic Features Readme

### PR DESCRIPTION
**Pull request by mrm9084**

### CODE REVIEW

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 1]
```diff
@@ -259,6 +259,104 @@ Options are available to customize how targeting evaluation is performed across
...
+  DiscountBanner:
+    Big:
+      size: 400
+      color: "#BBB"
+    Small:
+      size: 150
+      color: "#DDD"
+```
+
+A Dynamic Feature has two main parts the assigner and a list of variants. The assigner in this case `feature-management.DiscountBanner.assigner` is pointing the `FeatureFilter` `Microsoft.Targeting` that will evaluate the list of variants and choose which configuration-reference will be used. Each variant has; a name, configuration-reference, and assignment-parameters, and one of them needs to be set as the default variant. The name field is just meta data to reference the variant and just needs to be unique. The configuration-reference is a link to the configurations defined under `feature-variant` in the configuration file. **Note:** the colon notation is required when denoting parts of the configuration. If it is needed multiple colons can be used to reference sub configurations. i.e. `DiscountBanner:Big:size` would return the `Integer` 400.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:325]: Should have comma in first sentence.

`Microsoft.Targeting` in this case is an assigner not a feature filter.

Why a semi-colon in "has;" ?

I would change the ", and one of them needs to be set as the default variant." to it's own sentence. "One of the listed variants needs to be configured as the default variant for the feature."

The "just" usage in the "The name field is just..." sentence should be removed.

Calling the configuration reference a "link" might be confusing. Link usually refers to URL.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 2]
```diff
@@ -259,6 +259,104 @@ Options are available to customize how targeting evaluation is performed across
...
+
+The assignment-parameters work similarly to the other `FeatureFilters`. It can take in a `Map` which contains the evaluation criteria. In the case of `Microsoft.Targeting` it contains the `Audience` object. This enables custom assigners to be created.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:327]: Feature Assigners, not filters

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:327]: Perhaps switch the order of the last two sentences.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md (hunk 1 above)
Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:325]: It may be worth it to have a bulleted list of the properties to define their usage/requirements.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 3]
```diff
@@ -259,6 +259,104 @@ Options are available to customize how targeting evaluation is performed across
...
+
+The `Microsoft.Targeting` assigner uses three evaluation criteria to assign a variant; users, groups, and defaultRolloutPercentage. Assignment requires the definition of the `ITargetingContextAccessor` where a user defines how the user and group parameters are defined.
+
+If a users is listed in a variant then they are automatically given that variant. If they are in more than one variant they are returned to the first listed variant.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:331]: Instead of "given that variant" I think it would be better to say "assigned to that variant"

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 4]
```diff
@@ -259,6 +259,104 @@ Options are available to customize how targeting evaluation is performed across
...
+
+Groups have a rollout percentage. Using the user and group information users in groups are bucketed into parts of the rollout. This enables the a rollout where 10 of a group gets a variant, then when it increases to 50% that 10% still gets the variant with an additional 40% of the group now getting the variant. The rolloutPercentage for a group is shared across all variants that use that group.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:333]: "This enables the a rollout" should be "This enables a rollout"

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:333]: This paragraph is hard to follow. Also, don't we define targeting semantics elsewhere in such a way that we can reference them and just explain how targeting assignment builds on top of it?

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 5]
```diff
@@ -259,6 +259,104 @@ Options are available to customize how targeting evaluation is performed across
...
+
+If Variant A has Group "Dev" which is set to 50, and Variant B also has Group "Dev" which is also set to 50 then, if you need to increase Variant A to have 60% for "Dev" then Variant B needs to be decreased to 40 as the max rollout percentage is 100.
+
+In a different setup Variant A has "Dev" set to 50 and Variant B has it set to 40. Then Variant A is increased to 60, then 10% of Variant B will move to use Variant A while the last 10% will be added to Variant B. This 10% of users could have had either variant before depending on the defaultRolloutPercentage configured and the default variant if one isn't assigned.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:337]: This depends on ordering of the variants. I think we need an actual example with configuration to explain this concept.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 6]
```diff
@@ -259,6 +259,104 @@ Options are available to customize how targeting evaluation is performed across
...
+
+A variant as shown can be part of multiple groups, as can a user be in multiple groups. Being in multiple groups results in multiple evaluation chances per group.
+
+If a user isn't listed, isn't in a group, or wan't in the rollout percentage of a group then they will be evaluated as part of the default rollout percentage. The default rollout percentage works the same as the group rollout percentage. It's value across all variants can't add up to more than 100% and each user is bucketed into a different part of the rollout. Because of this all users need to have a user id defined otherwise they will all return the same variant.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:341]: wan't

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:341]: I think instead of "Because of this" it would be better to mention that the uniqueness of assignment in the group rollout comes from unique user id's up front rather than as a result of previous explanation.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 7]
```diff
@@ -259,6 +259,104 @@ Options are available to customize how targeting evaluation is performed across
...
+
+If the default rollout percentage doesn't add up to 100 then any un assigned user will be given the default variant.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:343]: No matter what, any unassigned user will be given the default variant.

Unassigned is one word.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md (hunk 2 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:327]: Not sure what you mean by the first comment. Assignment-parameters do work the same as feature filters as they are passed into the class and used to evaluate with Feature is returned.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 8]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
     }
 ```
 
+## Dynamic Features
+
+Using `FeatureManager`'s `getVariantAsync` method allows for returning of dynamic features based off a feature evaluator. Usage:
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:264]: Whats a "feature evaluator" I haven't heard any of us use that term.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 9]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+
+```java
+featureManager.getVariantAsync("DiscountBanner", DiscountBanner.class)
+```
+
+This example shows requesting a Dynamic Feature named "DiscountBanner", it requires the name of the Dynamic Feature and the type that needs to be returned.
```
[the remark this answers is not in the document]
Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:270]: Should bet two different sentences.

[the remark this answers is not in the document]
Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:270]: Perhaps it's worth noting that it returns an object of the type that is passed.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 10]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+          configuration-reference: "DiscountBanner:Big"
+          assignment-parameters:
+            ...
+        -
+          name: Small
+          default: false
+          configuration-reference: "DiscountBanner:Small"
+          assignment-parameters:
+            ...
+feature-variant:
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:294]: Do you think `feature-variant` or `feature-variants` is better considering we'll be defining all the feature variants here.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 11]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+  dynamic-features:
+    DiscountBanner:
+      assigner: "Microsoft.Targeting"
+      variants:
+        -
+          name: Big
+          default: true
+          configuration-reference: "DiscountBanner:Big"
+          assignment-parameters:
+            ...
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:287]: Do you mind putting some real assignment parameters, even if just targeting a single user.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 12]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+      color: "#BBB"
+    Small:
+      size: 150
+      color: "#DDD"
+```
+
+A Dynamic Feature has two main parts, the assigner and a list of variants. The assigner in this case `feature-management.DiscountBanner.assigner` which is pointing the `TargetingEvaluator` `Microsoft.Targeting` that will evaluate the list of variants and choose which configuration-reference will be used. Each variant has: a name, configuration-reference, and assignment-parameters. One of the listed variants needs to be configured as the default variant for the feature. The name field is meta data to reference the variant and just needs to be unique. The configuration-reference is a reference to the configurations defined under `feature-variant` in the configuration file. **Note:** the colon notation is required when denoting parts of the configuration. If it is needed multiple colons can be used to reference sub configurations. i.e. `DiscountBanner:Big:size` would return the `Integer` 400.
+
+The assignment-parameters takes in a `Map` which contains the evaluation criteria. This enables custom assigners to be created. In the case of `Microsoft.Targeting` it contains the `Audience` object.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:306]: Is the Audience defined somewhere to be linked to?

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 13]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+  DiscountBanner:
+    Big:
+      size: 400
+      color: "#BBB"
+    Small:
+      size: 150
+      color: "#DDD"
+```
+
+A Dynamic Feature has two main parts, the assigner and a list of variants. The assigner in this case `feature-management.DiscountBanner.assigner` which is pointing the `TargetingEvaluator` `Microsoft.Targeting` that will evaluate the list of variants and choose which configuration-reference will be used. Each variant has: a name, configuration-reference, and assignment-parameters. One of the listed variants needs to be configured as the default variant for the feature. The name field is meta data to reference the variant and just needs to be unique. The configuration-reference is a reference to the configurations defined under `feature-variant` in the configuration file. **Note:** the colon notation is required when denoting parts of the configuration. If it is needed multiple colons can be used to reference sub configurations. i.e. `DiscountBanner:Big:size` would return the `Integer` 400.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: The assigner in this case `feature-management.DiscountBanner.assigner` which is pointing the `TargetingEvaluator` `Microsoft.Targeting` that will evaluate the list of variants and choose which configuration-reference will be used.

Looks like it's missing some words?

The assigner used by a dynamic feature is specified by the `feature-management.DiscountBanner.assigner` property. In this example `Microsoft.Targeting` is specified which means that the `TargetingEvaluator`  will be used.

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: It's best to avoid fillers like "just" in technical writing.

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: metadata is one word.

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: I see assignment parameters is separated into a separate paragraph. Why not just do a bulleted list to describe all of their properties in their own bullet?

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 14]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+
+The assignment-parameters takes in a `Map` which contains the evaluation criteria. This enables custom assigners to be created. In the case of `Microsoft.Targeting` it contains the `Audience` object.
+
+#### Targeting Evaluator
+
+`TargetingEvaluator` is a built-in assigner, which can be enabled by creating the `Microsoft.Targeting` Bean.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:310]: terminology is confusing. it's called Evaluator but then we directly call it an assigner.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 15]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+public TargetingEvaluator targetingEvaluator(TargetingContextImpl context) {
+    return new TargetingEvaluator(context, new TargetingEvaluationOptions().setIgnoreCase(true));
+}
+```
+
+`TargetingEvaluator` extends `TargetingFilter` so both can be used at the same time.
+
+The `Microsoft.Targeting` assigner uses three evaluation criteria to assign a variant; users, groups, and defaultRolloutPercentage. Assignment requires the definition of the `ITargetingContextAccessor` where a user defines how the user and group parameters are defined.
+
+If a users is listed in a variant then they are automatically assigned to that variant. If they are in more than one variant they are returned to the first listed variant.
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:324]: I would give an example audience object with users, groups, and default rollout percentage before breaking down the categories.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md [hunk 16]
```diff
@@ -259,6 +259,171 @@ Options are available to customize how targeting evaluation is performed across
...
+  groups:
+    -
+        name: "Dev"
+        rolloutPercentage: 10
+  defaultRolloutPercentage: 0
+```
+
+to
+
+```yml
```
Comment (jimmyca15) [inline sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:339]: Is it necessary? I feel that saying to change the percentage from 10 to 50 have the same affect.

Comment (mrm9084) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:339]: Do you mean the move one to the other example?

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:339]: I just meant that copying the configuration block and only tweaking the percentage might just be as easy as mentioning that the percentage needs to be updated.

File: sdk/appconfiguration/azure-spring-cloud-feature-management/README.md (hunk 13 above)
Comment (mrm9084) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: The description seems a bit long for a bulleted list.

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: The description of each property is too long?

Comment (mrm9084) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: I just think bulleted lists are only good for short items.

Comment (jimmyca15) [reply sdk/appconfiguration/azure-spring-cloud-feature-management/README.md:304]: I don't disagree. I'm just wondering which property description isn't a short item?
